Steps Inputs 2

DanielDuchnicki opened the pull request.

DanielDuchnicki committed:
Adding new version of passing inputs into steps.

michalborowski commented on HotelReservation/HotelReservation/ConsolePrinter.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class ConsolePrinter
+    {
+        public virtual void Execute(string message)
```
Execute brzmi bardzo tajemniczo jak na klasę, która wiadomo co robi ;) Write/WriteLine bardziej się nadaje

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/IReservationStep.cs:
```diff
@@ -1,7 +1,9 @@
     public interface IReservationStep
     {
-        void Execute();
+        void Execute(ConsolePrinter consolePrinter, List<StepInput> stepData);
```
ConsolePrinter wygląda średnio jako argument metody. Czy każdy step wymaga consolePrintera? Obecnie tak, ale tylko przypadkiem. Nie chcesz tego pokazywać w publicznym interfejsie, który każdy step musi spełniać. Mamy factory, która może przekazać stepom współpracowników.

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Mail/SendingMailProcess.cs:
```diff
@@ -1,12 +1,12 @@
+        public void Execute(ConsolePrinter consolePrinter, List<StepInput> stepData)
         {
-            Console.WriteLine("----==== SENDING MAIL PROCESS ====----");
+            new ConsolePrinter().Execute("----==== SENDING MAIL PROCESS ====----");
```
coś tu jest nie tak, przekazujesz consolePrinter jako argument a tu jest new?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Reservation/ReservationStartProcess.cs:
```diff
@@ -0,0 +1,16 @@
+        string name;
+        public void Execute(ConsolePrinter consolePrinter, List<StepInput> stepData)
+        {
+            name = stepData.Find(stepInput => stepInput.Identifier == "name").Value;
```
Zastanów się, czy chcesz takie coś powielać w każdym stepie. Jeśli każdy step będzie miał coś w tym stylu, to może warto to uwspólnić. Innym pytaniem jest, czy step powinien to robić. Masz klasę executor, która w teorii powinna się zajmować odpaleniem stepów w poprawny sposób

DanielDuchnicki replied:
Póki co pozostało jako część stepów.

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,24 @@
+        public string QuestionText { get; private set; }
+        public dynamic Value { get; private set; }
+
+        public StepInput(Type type, string identifier, string questionText)
```
questionText - czy to nie jest odpowiedzialność innej warstwy?

DanielDuchnicki replied:
Usunięte

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Reservation/ReservationStartProcess.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class ReservationStartProcess : IReservationStep
+    {
+        string name;
```
Nie do końca czuję zastosowanie tego pola. Jeśli to jest sposób na zdefiniowanie inputów to coś tu nie gra. Trochę brakuje testów, które to tłumaczą. Przekazując tutaj stepy powinieneś napisać ładny zbiór testów na to, jak jest matchowany step, co, jeśli go nie ma itd.

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,24 @@
+        public Type InputType { get; private set; }
+        public string Identifier { get; private set; }
+        public string QuestionText { get; private set; }
+        public dynamic Value { get; private set; }
```
Dynamic bym póki co zostawił, Żaden z stepów nie operuje na czymś innym niż string

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,24 @@
+            QuestionText = questionText;
+        }
+
+        public void setValue(dynamic value)
```
Nazwa: Set zamiast set. 
Podejście: Stworzenie StepInput bez ustawionej wartości jest trochę niebezpieczne. Może warto ten problem rozbić? Identifier - nie do końca wiem, jak on ma działać. Proponuję skupić się bardziej na samym początku - tam gdzie ta klasa jest używana - w stepach. Coś tam nie do końca się spina i to wychodzi wyżej

DanielDuchnicki replied:
Fixed

DanielDuchnicki committed:
Adding some changes to the StepInputs functionality.

michalborowski commented on HotelReservation/ConsoleUserInterface/SystemMain.cs:
```diff
@@ -62,10 +61,10 @@ public static int Menu(HotelSystem hotelSystem, StepsExecutor stepsExecutor)
-                        Console.WriteLine(stepInput.QuestionText);
-                        stepInput.setValue(Console.ReadLine());
-                        stepsExecutor.ExecuteSteps(stepsList, new ConsolePrinter(), new List<StepInput> { stepInput });
+                        var stepInput = new StepInput(typeof(string), "name");
```
Skąd wiadomo jaki input?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/ReservationSteps/StepsExecutorTests.cs:
```diff
@@ -66,10 +64,28 @@ public void ShouldCreateStepsBasedOnProvidedListOfStepsTypes()
+        }
+
+        [Test]
+        public void ShouldThrowExceptionWhenNoDataFound()
```
:)

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepsExecutor.cs:
```diff
@@ -10,10 +10,24 @@ public StepsExecutor(StepFactory stepFactory)
         }
 
-        public void ExecuteSteps(List<ReservationStepType> reservationSteps, ConsolePrinter consolePrinter, List<StepInput> stepsData)
+        public List<StepInput> SetStepInputsValues(List<StepInput> stepInputs, List<StepInput> stepsData)
```
Wydaje mi się, że używasz klasy StepInput na dwa, zupełnie różne sposoby.

michalborowski replied:
Czy ta metoda nie powinna być prywatna?
Dodatkowo, zwraca ona listę, ale nie używasz tej zwracanej wartości.

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepsExecutor.cs:
```diff
@@ -10,10 +10,24 @@ public StepsExecutor(StepFactory stepFactory)
+            {
+                reservationStepInstance = _stepFactory.CreateInstance(reservationStep);
+                SetStepInputsValues(reservationStepInstance.GetStepInputs(), stepsData);
+                reservationStepInstance.Execute();
```
I wszystko to po nic, bo inputy nie są używane :D?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepsExecutor.cs:
```diff
@@ -10,10 +10,24 @@ public StepsExecutor(StepFactory stepFactory)
-                _stepFactory.CreateInstance(reservationStep).Execute(consolePrinter, stepsData);
+            {
+                reservationStepInstance = _stepFactory.CreateInstance(reservationStep);
+                SetStepInputsValues(reservationStepInstance.GetStepInputs(), stepsData);
```
Ustawianie wartości dla kroku jest bardzo nieprzejrzyste. Zmieniasz stan kroku przed wywołaniem Execute. Przez to miałeś problem z przetestowaniem przekazywania wartości. Wydaje mi się, że lepszym kierunkiem jest przekazanie wartości do kroku przy czyli Execute(stepInputs)

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Reservation/ReservationStartProcess.cs:
```diff
@@ -4,13 +4,25 @@ namespace HotelReservation.ReservationSteps.Reservation
 {
     public class ReservationStartProcess : IReservationStep
     {
-        string name;
```
Wydaje mi się, że w dalszym ciągu nie wymyśliłeś, jak chcesz używać inputów w stepie, a zacząłeś implementować ich obsługę wyżej w kodzie. Na przykład przekazujesz z samej góry typeof(string), co nie ma na nic wypływu. 

Najłatwiej, jak to sobie przegadamy w 4 oczy :) Ustawmy się w poniedziałek

DanielDuchnicki replied:
Fixed

DanielDuchnicki committed:
Adding 'QuestionIdentifier' enum with two questions: Name and EmailAddress

DanielDuchnicki committed:
- adding List<StepInput> as an argument to Execute method,
- changing StepExecutor class
- adding StepInput handling for ReservationStartProcess step

DanielDuchnicki committed:
Changes to SystemMain file and new "ProcessControler" class

DanielDuchnicki committed:
Removing unnecessary StepInput Type

DanielDuchnicki committed:
Adding ReserveHotelUseCase layer and its method to get hotels list.

DanielDuchnicki committed:
Adding method to get hotel reservation steps to the Reserve Hotel usecase.

DanielDuchnicki committed:
Adding CreateStepsInstances method and changing executing steps functionality

DanielDuchnicki committed:
Adding reading steps Inputs from step instances.

DanielDuchnicki committed:
Adding executing steps by Reserve Hotel Usecase.

DanielDuchnicki committed:
Adding general methods to execute steps and gather step inputs values.

DanielDuchnicki committed:
Adding functionality to remove duplicates from step input list and adding tests for SendingMailProcess.

DanielDuchnicki committed:
Changing assemblyInfo file for HotelReservation project and classes, methods, properties access modifiers.

michalborowski commented on HotelReservation/HotelReservation/Hotels/Hotel.cs:
```diff
@@ -5,10 +5,10 @@ namespace HotelReservation.Hotels
     {
-        public int? HotelId { get; }
-        public List<ReservationStepType> ReservationSteps { get; }
+        public int? HotelId { get; private set; }
```
private set can be removed (it'll stay private)

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ConsolePrinter.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace HotelReservation
+{
+    public class ConsolePrinter
```
Just a reminder: You shouldn't have any dependencies to console here. It's just something dummy to keep steps busy and make them testable. 

Ideally there should be ILogger or IMessageHandler interface in this class that doesn't say it's a console and is implemented by ConsolePrinter in ConsolUserInterface

michalborowski commented on HotelReservation/HotelReservation/Hotels/HotelSystem.cs:
```diff
@@ -5,24 +5,24 @@
 namespace HotelReservation.Hotels
 {
-    public class HotelSystem
+    internal class HotelSystem
```
When class is internal you can keep methods public

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Mail/SendingMailProcess.cs:
```diff
@@ -1,12 +1,25 @@
+            _consolePrinter = consolePrinter;
+            stepInputs = new List<StepInput> { name, email };
+        }
+        public List<StepInput> GetStepInputs()
```
Naming:
Worth to mention "Required" in names to show what these steps are

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/Mail/SendingMailProcess.cs:
```diff
@@ -1,12 +1,25 @@
     {
-        public void Execute()
+        private ConsolePrinter _consolePrinter;
+        private StepInput name = new StepInput(InputType.Name);
```
Do you need to keep it initialized and store it in private field? It can be tricky, because when someone will GetInputSteps() twice or more you'll return this exact instance each time. Change made to one of them will be visible in all others which leads to hart to find bugs. 
Nice to read about: 
* Immutability 
* Value vs reference types

What you truly need t keep is list of input types with two types there. Then I'd imagine something like:

public List<StepInput> GetStepInputs() => _requiredInputTypes.Select(type => new StepInput(type))

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class StepInput
+    {
+        public InputType Identifier { get; private set; }
```
Name: no longer identifier

DanielDuchnicki replied:
Fixed - changed to Type

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepFactory.cs:
```diff
@@ -1,33 +1,32 @@
-        public virtual IReservationStep CreateInstance(ReservationStepType reservationStep)
+        internal virtual IReservationStep CreateInstance(ReservationStepType reservationStep)
         {
             Func<IReservationStep> instance;
```
Name: instanceBuilder

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,21 @@
+        internal StepInput(InputType identifier)
+        {
+            Identifier = identifier;
+            Value = null;
```
No need for that

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReservationSteps/StepInput.cs:
```diff
@@ -0,0 +1,21 @@
+            Value = null;
+        }
+
+        public void SetValue(string value)
```
Why not making Value setter public?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            _systemInit.AddHotels(_hotelSystem);
+        }
+
+        public ReserveHotelUsecase()
```
You can't do that - your tests are not testing your production code :( You need one constructor. The problem you have with internal resources would be sorted out by DI. I think UseCaseFactory is something you're looking for now

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            _systemInit = new SystemInit();
+            _systemInit.AddHotels(_hotelSystem);
+        }
+        public ReadOnlyCollection<Hotel> GetHotels()
```
Use case should have one purpose - reserve a hotel. Getting a list of available hotels is another case of usage your system handles, another use case class

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            return _hotelSystem.GetHotels();
+        }
+
+        internal List<ReservationStepType> GetHotelReservationSteps(int hotelId)
```
This could be private

DanielDuchnicki replied:
Istnieją testy dla tej metody, więc chyba powinna pozostać internal?

michalborowski replied:
Testy powinny być wykonywane przez publiczny interfejs klasy. Jeśli testy nie używają logiki w ten sam sposób, jak kod produkcyjny (przez dostępny interfejs) są niewiele warte. Chcesz się przecież upewnić, że jak zawołasz metodę publiczną, to klasa odpowiednio zadziała. Popatrz na testy do tej klasy, przetestowałeś wszystkie metody oprócz publicznej - może być pusta, a wszystkie testy przejdą.

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            return new ReadOnlyCollection<StepInput>(stepInputs);
+        }
+
+        public ReadOnlyCollection<StepInput> GetRequiredStepInputsForHotelId(int hotelId)
```
This is another use case

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            return _hotelSystem.GetHotelReservationSteps(hotelId);
+        }
+
+        internal List<IReservationStep> CreateStepsInstances(List<ReservationStepType> reservationStepTypes)
```
Please keep the order of methods in class:
* Constructors
* Public 
* Internal
* Private

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            return GetStepsInputs(reservationSteps);
+        }
+
+        internal void ExecuteSteps(List<IReservationStep> reservationSteps, List<StepInput> stepsInputs)
```
This should be private

DanielDuchnicki replied:
Istnieją testy dla tej metody, więc chyba powinna pozostać internal?

michalborowski replied:
jw

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            _stepExecutor.ExecuteSteps(reservationSteps, stepsInputs);
+        }
+
+        public void ExecuteStepsForHotelId(int hotelId, List<StepInput> stepsInputs)
```
Why don't call it simpy ReserveHotel(...)

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+            return reservationSteps;
+        }
+
+        internal ReadOnlyCollection<StepInput> GetStepsInputs(List<IReservationStep> reservationSteps)
```
private

DanielDuchnicki replied:
Istnieją testy dla tej metody, więc chyba powinna pozostać internal?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/ReserveHotelUseCase.cs:
```diff
@@ -0,0 +1,83 @@
+                stepInputs.AddRange(reservationStep.GetStepInputs().
+                    Where(stepInputsList => stepInputs.FirstOrDefault(stepInput => stepInput.Identifier == stepInputsList.Identifier) == null));
+            }
+            return new ReadOnlyCollection<StepInput>(stepInputs);
```
Not sure I got it right but you can simplify things with Linq a bit better:
reservationSteps
              .SelectMany(step => step.GetStepInputs())
              .GroupBy(step => step.Identifier)
              .Select(grouping => grouping.First());

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/ReserveHotelUsecaseTests.cs:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        [Test]
+        public void GetHotelsShouldCallHotelSystemGetHotels()
```
Try to avoid writing test names based on method or classes names. This test name will become outdated if any of these names changes and you won't have any reference to it.

Instead, show the intention:
ShouldRetrieveAvailableHotels()

DanielDuchnicki replied:
Fixed

DanielDuchnicki committed:
Changing methods accessibility for internal classes and removing unnecessary private setters.

DanielDuchnicki committed:
Changing required step inputs handling in steps classes and method to get list of them.

DanielDuchnicki committed:
Some changes to methods and variables names and method to get required step inputs.

DanielDuchnicki committed:
Adding UseCaseFactory

DanielDuchnicki committed:
Changing order of methods and some methods accessibility.

michalborowski commented on HotelReservation/HotelReservation/UseCase.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace HotelReservation
+{
+    public abstract class UseCase
```
Klasa abstrakcyjna ma zastosowanie, gdy jakaś część logiki powinna być uwspólniona, a reszta dopisana w zależności od realizacji.
Gdy nie ma żadnej logiki - stosujemy interfejs.
Tutaj natomiast mamy interfejs, któremu trzeba dopisać metodę dodając jego kolejne implementacje. 
Miałoby to sens, gdyby każdy UseCase miał taką samą metodę: np. Execute

Proponuję usunąć, a w factory zwracać po prostu trzy oddzielne typy trzema metodami

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservation/UseCaseFactory.cs:
```diff
@@ -0,0 +1,42 @@
+
+        public UseCaseFactory()
+        {
+            _hotelSystem = new HotelSystem();
```
Dlaczego tu robisz new, a dla usecase już nie?

DanielDuchnicki replied:
Po dyskusji - zmiana na razie niepotrzebna

michalborowski commented on HotelReservation/HotelReservationTests/GetHotelRequiredStepInputsUseCaseTests.cs:
```diff
@@ -0,0 +1,45 @@
+namespace HotelReservationTests
+{
+    [TestFixture]
+    class GetHotelRequiredStepInputsUseCaseTests
```
Publiczna metoda nie jest przetestowana

DanielDuchnicki replied:
Fixed?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/GetHotelsUseCaseTests.cs:
```diff
@@ -0,0 +1,30 @@
+        }
+
+        [Test]
+        public void ShouldRetrieveAvailableHotels()
```
Nie testujesz co jest zwracane

DanielDuchnicki replied:
Fixed - dodany test

michalborowski commented on HotelReservation/HotelReservationTests/ReservationSteps/Mail/SendingMailProcessTests.cs:
```diff
@@ -14,31 +15,56 @@ class SendingMailProcessTests
+
+            _subject.Execute(new List<StepInput> { _name, _mail });
+
+            A.CallTo(() => _consolePrinterDouble.Write(nameValue)).MustHaveHappened();
```
To powinny być dwa testy. Kiedy ten test nie przejdzie nie wiesz od razu, którego asserta to dotyczy.

DanielDuchnicki replied:
Fixed

DanielDuchnicki committed:
Creating separate tests for name and mail values.

DanielDuchnicki committed:
Adding test to check what is returned by GetHotels UseCase

DanielDuchnicki committed:
Adding tests for public method

DanielDuchnicki committed:
Adding tests for public method (GetHotelRequiredStepInputs UseCase)

DanielDuchnicki committed:
Changes to the UseCase factory and its tests. Removing unnecessary interface and enum.

michalborowski commented on HotelReservation/HotelReservationTests/GetHotelsUseCaseTests.cs:
```diff
@@ -0,0 +1,40 @@
+        public void ShouldReturnCurrentHotelsInHotelSystem()
+        {
+            var dummyHotels = new ReadOnlyCollection<Hotel>(new List<Hotel>());
+            A.CallTo(() => _hotelSystemDouble.GetHotels()).Returns(dummyHotels);
```
Czy ten test nie zawiera poprzedniego?

DanielDuchnicki replied:
Prawda. Fixed

DanielDuchnicki committed:
Changes to UseCases tests

DanielDuchnicki committed:
Changes to UseCases' methods accessibility

DanielDuchnicki committed:
Some minor changes to tests, systemMain and StepInput

michalborowski commented on HotelReservation/HotelReservation/GetHotelRequiredStepInputsUseCase.cs:
```diff
@@ -24,22 +24,22 @@ public ReadOnlyCollection<StepInput> GetRequiredStepInputsForHotelId(int hotelId
         }
 
-        internal List<IReservationStep> CreateStepsInstances(List<ReservationStepType> reservationStepTypes)
+        private List<IReservationStep> CreateStepsInstances(List<ReservationStepType> reservationStepTypes)
```
Zauważ, że cała ta metoda jest w 2 use caseach i do tego jest zupełnie identyczna. Każdy use case z inputami będzie musiał posiadać taką metodę. Co więcej testowanie foreacha nie jest takie łatwe. Może warto mieć całą tę metodę w oddzielnej klasie i testować tylko raz?

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/ReservationSteps/StepsExecutorTests.cs:
```diff
@@ -10,7 +10,7 @@
 namespace HotelReservationTests.ReservationSteps
 {
     [TestFixture]
```
Właśnie zauważyłem, że masz tutaj testy dla każdego typu stepów. Nie jest to konieczne. Ta klasa pracuje z IReservationStep właśnie po to, żeby nie przejmować się konkretnymi realizacjami tego interfejsu

DanielDuchnicki replied:
Fixed

michalborowski commented on HotelReservation/HotelReservationTests/ReserveHotelUsecaseTests.cs:
```diff
@@ -26,88 +25,58 @@ public void BeforeTest()
         [Test]
         public void ShouldExecuteStepsForProvidedHotelId()
         {
             var dummyHotelId = 1;
```
W takim przypadku używamy const (masz tak w kilku miejscach)

DanielDuchnicki replied:
Fixed

michalborowski reviewed (approved):
Małe zmiany, możesz mergować od razu po wprowadzeniu.

DanielDuchnicki committed:
Removing unnecessary (redundant) test of GetHotels method.

DanielDuchnicki committed:
Removing unnecessary (redundant) tests for StepsExecutor

DanielDuchnicki committed:
Adding const inst instead of var for dummyId in tests

DanielDuchnicki committed:
Moving CreateStepInstance method to new class and adding tests for it.
